transports/janus_mqtt.c

```diff
@@ -0,0 +1,648 @@
+					goto error;
+				}
+				ctx->admin.publish.topic = g_strdup(topic_item->value);
+
+				janus_config_item *qos_item = janus_config_get_item_drilldown(config, "admin.publish", "qos");
+				ctx->admin.publish.qos = (qos_item && qos_item->value) ? atoi(qos_item->value) : 1;
+			}
+		}
+		else {
+			janus_mqtt_admin_api_enabled_ = FALSE;
+			ctx->admin.subscribe.topic = NULL;
+			ctx->admin.publish.topic = NULL;
+		}
+	}
+
+	if(!janus_mqtt_api_enabled_ && !janus_mqtt_admin_api_enabled_) {
+		JANUS_LOG(LOG_WARN, "MQTT support disabled for both Janus and Admin API, giving up\n");
+		goto error;
+	}
+
+	/* Creating a client */
+	if(MQTTAsync_create(
+			&ctx->client,
+			url,
+			client_id,
+			MQTTCLIENT_PERSISTENCE_NONE,
+			NULL) != MQTTASYNC_SUCCESS) {
+		JANUS_LOG(LOG_FATAL, "Can't connect to MQTT broker: error creating client...\n");
+		goto error;
+	}
+	if(MQTTAsync_setCallbacks(
+			ctx->client,
+			ctx,
+			janus_mqtt_client_connection_lost,
+			janus_mqtt_client_message_arrived,
+			janus_mqtt_client_delivery_complete) != MQTTASYNC_SUCCESS) {
+		JANUS_LOG(LOG_FATAL, "Can't connect to MQTT broker: error setting up callbacks...\n");
+		goto error;
+	}
+
+	/* Connecting to the broker */
+	int rc = janus_mqtt_client_connect(ctx);
+	if(rc != MQTTASYNC_SUCCESS) {
+		JANUS_LOG(LOG_FATAL, "Can't connect to MQTT broker, return code: %d\n", rc);
+		goto error;
+	}
+
+	return 0;
+
+error:
+	/* If we got here, something went wrong */
+	janus_mqtt_client_destroy_context(ctx);
+	g_free((char *)url);
+	g_free((char *)client_id);
+	g_free((char *)username);
+	g_free((char *)password);
+	g_free(config);
+
+	return -1;
+}
+
+void janus_mqtt_destroy(void) {
+	JANUS_LOG(LOG_INFO, "Disconnecting MQTT client...\n");
+
+	janus_mqtt_client_disconnect(context_);
+}
+
+int janus_mqtt_get_api_compatibility(void) {
+	return JANUS_TRANSPORT_API_VERSION;
+}
+
+int janus_mqtt_get_version(void) {
+	return JANUS_MQTT_VERSION;
+}
+
+const char *janus_mqtt_get_version_string(void) {
+	return JANUS_MQTT_VERSION_STRING;
+}
+
+const char *janus_mqtt_get_description(void) {
+	return JANUS_MQTT_DESCRIPTION;
+}
+
+const char *janus_mqtt_get_name(void) {
+	return JANUS_MQTT_NAME;
+}
+
+const char *janus_mqtt_get_author(void) {
+	return JANUS_MQTT_AUTHOR;
+}
+
+const char *janus_mqtt_get_package(void) {
+	return JANUS_MQTT_PACKAGE;
+}
+
+gboolean janus_mqtt_is_janus_api_enabled(void) {
+	return janus_mqtt_api_enabled_;
+}
+
+gboolean janus_mqtt_is_admin_api_enabled(void) {
+	return janus_mqtt_admin_api_enabled_;
+}
+
+int janus_mqtt_send_message(void *context, void *request_id, gboolean admin, json_t *message) {
+	if(message == NULL) {
+		return -1;
+	}
+	if(context == NULL) {
+		json_decref(message);
+		return -1;
+	}
+
+	janus_mqtt_context *ctx = (janus_mqtt_context *)context;
+	char *payload = json_dumps(message, json_format_);
+	json_decref(message);
+
+	JANUS_LOG(LOG_HUGE, "Sending %s API message via MQTT: %s\n", admin ? "admin" : "Janus", payload);
```

Comment:
I use the async API of client library. The messages are internally being queued by the library, and then being dispatched by dedicated thread.